lib/pghero/methods/tables.rb

```diff
@@ -1,67 +1,287 @@
+                      sum(heap_blks_read) AS heap_blks_read
+                    FROM
+                      pg_statio_user_tables
+                  )
+                  worker_heap_blks_hit_read $cmd$)
+            )
+            SELECT
+              sum(heap_blks_hit) / nullif(sum(heap_blks_hit) + sum(heap_blks_read), 0) AS rate
+            FROM
+              (
+                SELECT
+                  heap_blks_hit,
+                  heap_blks_read
+                FROM
+                  pg_statio_user_tables
+                UNION ALL
+                SELECT
+                  workers.heap_blks_hit,
+                  workers.heap_blks_read
+                FROM
+                  worker_heap_blks_hit_read_json
+                CROSS JOIN
+                  json_to_recordset(worker_heap_blks_hit_read_json.result) AS workers("heap_blks_hit" int, "heap_blks_read" int)
+              ) AS cluster_heap_blks_hit_read
+          SQL
+          )
+        end
       end
 
       def table_caching
-        select_all <<-SQL
-          SELECT
-            schemaname AS schema,
-            relname AS table,
-            CASE WHEN heap_blks_hit + heap_blks_read = 0 THEN
-              0
-            ELSE
-              ROUND(1.0 * heap_blks_hit / (heap_blks_hit + heap_blks_read), 2)
-            END AS hit_rate
-          FROM
-            pg_statio_user_tables
-          ORDER BY
-            2 DESC, 1
-        SQL
+        if !citus_enabled?
+          select_all <<-SQL
+            SELECT
+              schemaname AS schema,
+              relname AS table,
+              CASE WHEN heap_blks_hit + heap_blks_read = 0 THEN
+                0
+              ELSE
+                ROUND(1.0 * heap_blks_hit / (heap_blks_hit + heap_blks_read), 2)
+              END AS hit_rate
+            FROM
+              pg_statio_user_tables
+            ORDER BY
+              2 DESC, 1
+          SQL
+        else
+          select_all <<-SQL
+            WITH dist_tables_hit_rate AS (
+              WITH placements_heap_blks_hit_read AS (
+                SELECT
+                  logicalrelid,
+                  (run_command_on_placements(logicalrelid, $$
+                    SELECT
+                      heap_blks_hit AS hit
+                    FROM
+                      pg_statio_user_tables
+                    WHERE
+                      relname = '%s' $$ )).result::int AS heap_blks_hit,
+                  (run_command_on_placements(logicalrelid, $$
+                    SELECT
+                      heap_blks_read
+                    FROM
+                      pg_statio_user_tables
+                    WHERE
+                      relname = '%s' $$ )).result::int AS heap_blks_read
+                FROM
+                  pg_dist_partition
+              )
+              SELECT
+                logicalrelid::name as dist_table_name,
+                CASE WHEN sum(heap_blks_hit + heap_blks_read) = 0 THEN
```

Comment:
I'd say lets leave hits and reads separate and dont calculate hit_rate here. Instead UNION this table with the stats table of non-distributed table and calculate hit_rate of all of them together. This would simplify the code and save us from this CASE statement here.

Reply:
Yes, much better in this way 👍